File: mac/Marquee/Connection/ServerProbe.swift
import Foundation

/// `GET /api/v1/server-info`, the public discovery endpoint (Docs/API_V1_CORE.md).
struct ServerInfo: Codable, Equatable, Hashable, Sendable {
    /// Always `"marquee"`; anything else isn't one of ours.
    let app: String
    let apiVersion: Int
    /// The server's package.json version, e.g. `"0.22.0"`.
    let version: String
    /// Nil when the server can't reach its database (`status == "degraded"`).
    let setupComplete: Bool?
    let status: String

    static let supportedAPIVersion = 1
    /// The first server release with the v1 API — what "Update required" asks for.
    static let minimumServerVersion = "0.22.0"

    var isDegraded: Bool { status == "degraded" }

    init(app: String = "marquee", apiVersion: Int = supportedAPIVersion, version: String, setupComplete: Bool?, status: String = "ok") {
        self.app = app
        self.apiVersion = apiVersion
        self.version = version
        self.setupComplete = setupComplete
        self.status = status
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        app = try container.decode(String.self, forKey: .app)
        apiVersion = try container.decode(Int.self, forKey: .apiVersion)
        version = try container.decodeIfPresent(String.self, forKey: .version) ?? "unknown"
        setupComplete = try container.decodeIfPresent(Bool.self, forKey: .setupComplete)
        status = try container.decodeIfPresent(String.self, forKey: .status) ?? "ok"
    }
}

/// Why a server couldn't be reached — each gets its own explanation in the UI.
enum UnreachableReason: Equatable, Hashable, Sendable {
    /// Connection refused: the computer is there, but nothing listens on that port.
    case refused
    /// Timed out, or the host is down.
    case noResponse
    /// The host name didn't resolve (e.g. `tower.local` isn't on this network).
    case unknownHost
    /// macOS blocked the connection: Marquee lacks Local Network access.
    case localNetworkDenied
    /// Anything else (TLS failure, …), with the system's message.
    case failed(String)
}

/// What answered at an address.
enum ProbeOutcome: Equatable, Hashable, Sendable {
    /// A Marquee server with the API version this app speaks.
    case marquee(ServerInfo)
    /// A Marquee server older than 0.22.0: no v1 API, so `/api/v1/server-info`
    /// redirected to the `/login` HTML page.
    case legacy
    /// A Marquee server speaking an API version this app doesn't know.
    case incompatible(ServerInfo)
    /// Something answered, but it isn't Marquee.
    case notMarquee
    case unreachable(UnreachableReason)

    var serverInfo: ServerInfo? {
        if case let .marquee(info) = self { return info }
        return nil
    }

    /// The inline error for manual entry and saved-server checks; nil for a usable server.
    func problemMessage(for address: ServerAddress) -> String? {
        let name = address.displayName
        switch self {
        case .marquee:
            return nil
        case .legacy:
            return "Found Marquee at \(name), but the server needs updating to \(ServerInfo.minimumServerVersion) or later to work with the Mac app."
        case let .incompatible(info):
            return "The server at \(name) runs Marquee \(info.version), which is newer than this app supports. Update Marquee for Mac."
        case .notMarquee:
            return "\(name) responded, but it isn't a Marquee server. Check the address and port."
        case let .unreachable(reason):
            switch reason {
            case .refused:
                return "Nothing is answering on port \(address.effectivePort) at \(address.host). Check the port and that Marquee is running."
            case .noResponse:
                return "Couldn't reach \(name). Check the address and that the server is on."
            case .unknownHost:
                return "Couldn't find \(address.host) on your network. Try its IP address instead."
            case .localNetworkDenied:
                return "Marquee doesn't have Local Network access. Turn it on in System Settings › Privacy & Security › Local Network."
            case let .failed(message):
                return "Couldn't connect to \(name): \(message)"
            }
        }
    }
}

/// Identifies whatever is listening at an address by asking for
/// `/api/v1/server-info`, the same check discovery runs on every open port.
enum ServerProbe {
    static let infoPath = "/api/v1/server-info"

    /// Ephemeral and cookie-less, so a web session cookie never changes what
    /// the server answers, and nothing is cached between probes.
    private static let session: URLSession = {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 6
        configuration.timeoutIntervalForResource = 10
        configuration.waitsForConnectivity = false
        configuration.httpShouldSetCookies = false
        configuration.httpCookieAcceptPolicy = .never
        configuration.urlCache = nil
        configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
        configuration.httpAdditionalHeaders = ["User-Agent": "Marquee-macOS/\(AppInfo.version)"]
        return URLSession(configuration: configuration)
    }()

    /// Full check for a single address (manual entry, saved server): a TCP
    /// connect first, which tells "refused", "no response" and "Local Network
    /// denied" apart, then the HTTP request.
    @concurrent
    static func probe(
        _ address: ServerAddress,
        connectTimeout: Duration = .seconds(4),
        requestTimeout: TimeInterval = 6
    ) async -> ProbeOutcome {
        switch await TCPProbe.check(host: address.host, port: address.effectivePort, timeout: connectTimeout, ipv4Only: false) {
        case .open:
            return await fetchInfo(address, timeout: requestTimeout)
        case .refused:
            return .unreachable(.refused)
        case .noResponse:
            return .unreachable(.noResponse)
        case .unknownHost:
            return .unreachable(.unknownHost)
        case .localNetworkDenied:
            return .unreachable(.localNetworkDenied)
        }
    }

    /// Just the HTTP half, for callers that already know the port is open.
    @concurrent
    static func fetchInfo(_ address: ServerAddress, timeout: TimeInterval) async -> ProbeOutcome {
        guard let url = URL(string: address.baseURLString + infoPath) else { return .notMarquee }
        var request = URLRequest(url: url, timeoutInterval: timeout)
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else { return .notMarquee }
            return classify(
                statusCode: http.statusCode,
                contentType: http.value(forHTTPHeaderField: "Content-Type"),
                apiHeader: http.value(forHTTPHeaderField: APIClient.apiHeader),
                body: data
            )
        } catch {
            return .unreachable(unreachableReason(for: error))
        }
    }

    /// Pure classification of a server-info response, so it's unit-testable
    /// from canned bodies.
    static func classify(statusCode: Int, contentType: String?, apiHeader: String?, body: Data) -> ProbeOutcome {
        if (200..<300).contains(statusCode),
           let info = try? JSONDecoder().decode(ServerInfo.self, from: body),
           info.app == "marquee" {
            return info.apiVersion == ServerInfo.supportedAPIVersion ? .marquee(info) : .incompatible(info)
        }
        if isLegacyMarqueePage(contentType: contentType, body: body) {
            return .legacy
        }
        if apiHeader != nil {
            // A v1 server that failed to answer server-info (it's meant to
            // return 200 even when degraded) — it's ours, but not usable now.
            return .unreachable(.failed("the server returned an error (\(statusCode))."))
        }
        return .notMarquee
    }

    /// Legacy servers answer every unknown route with the web app's HTML,
    /// whose root layout sets `<title>Marquee</title>` (app/layout.tsx).
    static func isLegacyMarqueePage(contentType: String?, body: Data) -> Bool {
        let head = String(decoding: body.prefix(256_000), as: UTF8.self)
        let looksLikeHTML = contentType?.lowercased().contains("html") == true
            || head.range(of: "<html", options: .caseInsensitive) != nil
        guard looksLikeHTML else { return false }
        return head.range(of: #"<title[^>]*>\s*Marquee\s*</title>"#, options: [.regularExpression, .caseInsensitive]) != nil
    }

    static func unreachableReason(for error: Error) -> UnreachableReason {
        guard let urlError = error as? URLError else {
            return .failed(error.localizedDescription)
        }
        switch urlError.code {
        case .cannotConnectToHost:
            return .refused
        case .timedOut, .networkConnectionLost, .notConnectedToInternet:
            return .noResponse
        case .cannotFindHost, .dnsLookupFailed:
            return .unknownHost
        default:
            return .failed(urlError.localizedDescription)
        }
    }
}

File: windows/Marquee.Core/Connection/ServerProbe.cs
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;
using Marquee.Core.Api;
using Marquee.Core.Models;

namespace Marquee.Core.Connection;

public enum UnreachableReasonKind
{
    /// <summary>Connection refused: the computer is there, but nothing listens on that port.</summary>
    Refused,

    /// <summary>Timed out, or the host is down.</summary>
    NoResponse,

    /// <summary>The host name didn't resolve (e.g. <c>tower.local</c> isn't on this network).</summary>
    UnknownHost,

    /// <summary>Windows blocked the connection: the app lacks local network access.</summary>
    LocalNetworkDenied,

    /// <summary>Anything else (TLS failure, ...), with the system's message.</summary>
    Failed,
}

/// <summary>Why a server couldn't be reached; each gets its own explanation in the UI.</summary>
public sealed record UnreachableReason(UnreachableReasonKind Kind, string? Detail = null)
{
    public static readonly UnreachableReason Refused = new(UnreachableReasonKind.Refused);
    public static readonly UnreachableReason NoResponse = new(UnreachableReasonKind.NoResponse);
    public static readonly UnreachableReason UnknownHost = new(UnreachableReasonKind.UnknownHost);
    public static readonly UnreachableReason LocalNetworkDenied = new(UnreachableReasonKind.LocalNetworkDenied);

    public static UnreachableReason Failed(string message) => new(UnreachableReasonKind.Failed, message);
}

/// <summary>What answered at an address.</summary>
public abstract record ProbeOutcome
{
    private ProbeOutcome()
    {
    }

    /// <summary>A Marquee server with the API version this app speaks.</summary>
    public sealed record Marquee(ServerInfo Info) : ProbeOutcome;

    /// <summary>
    /// A Marquee server older than 0.22.0: no v1 API, so <c>/api/v1/server-info</c>
    /// redirected to the <c>/login</c> HTML page.
    /// </summary>
    public sealed record Legacy : ProbeOutcome;

    /// <summary>A Marquee server speaking an API version this app doesn't know.</summary>
    public sealed record Incompatible(ServerInfo Info) : ProbeOutcome;

    /// <summary>Something answered, but it isn't Marquee.</summary>
    public sealed record NotMarquee : ProbeOutcome;

    public sealed record Unreachable(UnreachableReason Reason) : ProbeOutcome;

    public ServerInfo? ServerInfo => this is Marquee marquee ? marquee.Info : null;

    /// <summary>The inline error for manual entry and saved-server checks; null for a usable server.</summary>
    public string? ProblemMessage(ServerAddress address)
    {
        var name = address.DisplayName;
        return this switch
        {
            Marquee => null,
            Legacy => $"Found Marquee at {name}, but the server needs updating to {Models.ServerInfo.MinimumServerVersion} or later to work with the Windows app.",
            Incompatible incompatible => $"The server at {name} runs Marquee {incompatible.Info.Version}, which is newer than this app supports. Update Marquee for Windows.",
            NotMarquee => $"{name} responded, but it isn't a Marquee server. Check the address and port.",
            Unreachable unreachable => unreachable.Reason.Kind switch
            {
                UnreachableReasonKind.Refused => $"Nothing is answering on port {address.EffectivePort} at {address.Host}. Check the port and that Marquee is running.",
                UnreachableReasonKind.NoResponse => $"Couldn't reach {name}. Check the address and that the server is on.",
                UnreachableReasonKind.UnknownHost => $"Couldn't find {address.Host} on your network. Try its IP address instead.",
                UnreachableReasonKind.LocalNetworkDenied => "Windows blocked Marquee from reaching your local network. Check the app's network permissions in Settings.",
                _ => $"Couldn't connect to {name}: {unreachable.Reason.Detail}",
            },
            _ => null,
        };
    }
}

/// <summary>
/// Identifies whatever is listening at an address by asking for
/// <c>/api/v1/server-info</c>, the same check discovery runs on every open port.
/// </summary>
public static class ServerProbe
{
    public const string InfoPath = "/server-info";
    public const string LegacyLoginPath = "/login";
    public static readonly TimeSpan RequestTimeout = TimeSpan.FromSeconds(6);

    private static readonly Regex MarqueeTitle = new(@"<title[^>]*>\s*Marquee\s*</title>", RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);

    /// <summary>
    /// Full check for a single address (manual entry, saved server). Sends
    /// no credentials: the answer must not depend on who is asking.
    /// </summary>
    /// <param name="handler">Tests pass a stub; the app leaves it null for the shared pool.</param>
    public static async Task<ProbeOutcome> ProbeAsync(
        ServerAddress address,
        HttpMessageHandler? handler = null,
        TimeSpan? timeout = null,
        CancellationToken ct = default)
    {
        var client = new ApiClient(address.BaseUrl, handler: handler);
        var requestTimeout = timeout ?? RequestTimeout;

        ApiClient.RawResponse raw;
        try
        {
            raw = await client.SendRawAsync(HttpMethod.Get, InfoPath, timeout: requestTimeout, ct: ct).ConfigureAwait(false);
        }
        catch (ApiException error)
        {
            if (error.IsCancellation)
            {
                ct.ThrowIfCancellationRequested();
            }
            return new ProbeOutcome.Unreachable(UnreachableReasonFor(error));
        }

        // A legacy server answers with a redirect to its HTML login page. The
        // transport never follows redirects, so the probe takes that one hop
        // itself, and only on the same host: it is the page's <title> that
        // proves this is Marquee, not the redirect (plenty of apps have a
        // /login).
        if (raw.IsRedirect && LoginPageUrl(address, raw.Location) is { } loginPage)
        {
            try
            {
                var page = await client.SendRawAsync(HttpMethod.Get, loginPage, timeout: requestTimeout, ct: ct).ConfigureAwait(false);
                return IsLegacyMarqueePage(page.ContentType, page.Body) ? new ProbeOutcome.Legacy() : new ProbeOutcome.NotMarquee();
            }
            catch (ApiException error)
            {
                if (error.IsCancellation)
                {
                    ct.ThrowIfCancellationRequested();
                }
                return new ProbeOutcome.Unreachable(UnreachableReasonFor(error));
            }
        }

        return Classify(raw.StatusCode, raw.ContentType, raw.HasApiHeader, raw.Body);
    }

    /// <summary>
    /// Pure classification of a server-info response, so it's unit-testable
    /// from canned bodies.
    /// </summary>
    public static ProbeOutcome Classify(int statusCode, string? contentType, bool hasApiHeader, byte[] body)
    {
        if (statusCode is >= 200 and < 300 && DecodeInfo(body) is { IsMarquee: true } info)
        {
            return info.IsSupported ? new ProbeOutcome.Marquee(info) : new ProbeOutcome.Incompatible(info);
        }
        if (IsLegacyMarqueePage(contentType, body))
        {
            return new ProbeOutcome.Legacy();
        }
        if (hasApiHeader)
        {
            // A v1 server that failed to answer server-info (it's meant to
            // return 200 even when degraded): it's ours, but not usable now.
            return new ProbeOutcome.Unreachable(UnreachableReason.Failed($"the server returned an error ({statusCode})."));
        }
        return new ProbeOutcome.NotMarquee();
    }

    /// <summary>
    /// Legacy servers answer every unknown route with the web app's HTML,
    /// whose root layout sets <c>&lt;title&gt;Marquee&lt;/title&gt;</c> (app/layout.tsx).
    /// </summary>
    public static bool IsLegacyMarqueePage(string? contentType, byte[] body)
    {
        var head = Encoding.UTF8.GetString(body, 0, Math.Min(body.Length, 256_000));
        var looksLikeHtml = contentType?.Contains("html", StringComparison.OrdinalIgnoreCase) == true
            || head.Contains("<html", StringComparison.OrdinalIgnoreCase);
        return looksLikeHtml && MarqueeTitle.IsMatch(head);
    }

    public static UnreachableReason UnreachableReasonFor(ApiException error)
    {
        if (error.Kind != ApiErrorKind.Network)
        {
            return UnreachableReason.Failed(error.Message);
        }
        return error.Failure switch
        {
            NetworkFailure.Refused => UnreachableReason.Refused,
            NetworkFailure.Timeout or NetworkFailure.ConnectionLost => UnreachableReason.NoResponse,
            NetworkFailure.UnknownHost => UnreachableReason.UnknownHost,
            NetworkFailure.AccessDenied => UnreachableReason.LocalNetworkDenied,
            _ => UnreachableReason.Failed(error.Message),
        };
    }

    /// <summary>The redirect target when it is this server's own <c>/login</c>; null for anywhere else.</summary>
    private static Uri? LoginPageUrl(ServerAddress address, string? location)
    {
        if (string.IsNullOrEmpty(location) || !Uri.TryCreate(address.BaseUrl, location, out var target))
        {
            return null;
        }
        var sameOrigin = string.Equals(
            target.GetLeftPart(UriPartial.Authority),
            address.BaseUrl.GetLeftPart(UriPartial.Authority),
            StringComparison.OrdinalIgnoreCase);
        return sameOrigin && string.Equals(target.AbsolutePath, LegacyLoginPath, StringComparison.OrdinalIgnoreCase) ? target : null;
    }

    private static ServerInfo? DecodeInfo(byte[] body)
    {
        try
        {
            return JsonSerializer.Deserialize<ServerInfo>(body, Json.Options);
        }
        catch (JsonException)
        {
            return null;
        }
    }
}
